### ISSUE TITLE
When normalizing targets (groups, party, resource, ven id) combine like targets into a single dict

### ISSUE THREAD

**Issue by jmuraca** [OpenLEADR/openleadr-python#28]
Currently each target has it's own key/value pair.  Target output for test harness test E1_0300_TH_VTN

`[{'group_id': 'NoMatchGroupID'}, {'group_id': 'NoMatch_1_GroupID'}, {'resource_id': 'NoMatchResourceID'}, {'resource_id': 'NoMatch_1_ResourceID'}, {'ven_id': 'NoMatchVenID'}, {'party_id': 'NoMatchParty_ID'}, {'party_id': 'NoMatch_1_Party_ID'}]`

This is frustrating as we need to check every key in th elist.

Suggestion to combine all like targets into a single key

`{group_id: {NoMatchGroupID, NoMatch_1_GroupID}, resource_id: {NoMatchResourceID, NoMatch_1_ResourceID}, ven_id: {NoMatchVenID}, party_id: {NoMatchParty_ID, NoMatcNoMatch_1_Party_ID_1_GroupID}}`

**Comment by stan-janssen** [comment]
A previous version of the Target normalizer would indeed convert the target to something like you propose:

```python
{
  'group_ids': ['group_one', 'group_two'], 
  'resource_ids': ['resource_one', 'resource_two']
}
```
The problem with this approach was that is was too different from the original EiTarget object.

In the new form, you could at least use this one-liner to get all the targets of the same type:

```python
group_ids = [value for key, value in targets.items() if key == 'group_id']
```

Perhaps I could add a new utility function that will do the conversion; something like this:

```python
def group_targets_by_type(list_of_targets):
    targets_by_type = {}
    for target in list_of_targets:
        for key, value in target.items():
            if key not in targets_by_type:
                targets_by_type[key] = []
            targets_by_type[key].append(value)
    return targets_by_type
```

I'll think about whether this form does make more sense in total, but I remember having used it before and deciding against it some time ago.

**Comment by stan-janssen** [comment]
I'm thinking about the following solution:

- Keep the 'targets' element of the event dictionary as-is, containing a list of EiTarget dicts.
- Add a second key 'targets_by_type' that contains a dict of targets grouped by type. An event dict would then look like this:

```python3
{
  'event_id': 'event001',
  # ...
  'targets': [{'group_id': 'group01'},
              {'group_id': 'group02'},
              {'resource_id': 'resource01'},
              {'resource_id': 'resource02'}],
  'targets_by_type': {'group_id': ['group01', 'group02'],
                      'resource_id': ['resource01', 'resource02']}
}
```

- When creating an event using the dict-constructor, either form should be acceptable. I will add a check to the preflight function to check that only one of both is given, or that both are equivalent.
- When creating an event using the object-constructor (`objects.Event`), either form should be acceptable. During `__post_init__` of that object, a check will be placed to figure out the correct form. It also copies the targets to the other form so that both are accessible equally.

When receiving the event, you can access the targets either way.

I will work up this solution sometime in the coming days. What do you think of this solution?

**Comment by jmuraca** [comment]
Having a separate list is good as I hadn't considered the depth of this change on templates etc.
Having targets listed like this would need to be split later for event parsing and actions, so this is helpful!
I'll test later today

**Comment by stan-janssen** [comment]
I've released this in version 0.5.11 just now. Thanks for the suggestion, it's indeed better this way.
